loader/src/ui/nodes/TextAreaV2.cpp

```diff
@@ -0,0 +1,377 @@
+IMPL_GETTER_SETTER(float, float, textScale, TextScale)
+IMPL_GETTER_SETTER(float, float, linePadding, LinePadding)
+IMPL_GETTER_SETTER(float, float, minWidth, MinWidth)
+IMPL_GETTER_SETTER(float, float, maxWidth, MaxWidth)
+IMPL_GETTER_SETTER(float, float, minHeight, MinHeight)
+IMPL_GETTER_SETTER(float, float, maxHeight, MaxHeight)
+IMPL_GETTER_SETTER(bool, bool, hyphenate, Hyphenate)
+IMPL_GETTER_SETTER(bool, bool, ellipsis, Ellipsis)
+IMPL_GETTER(bool, deferUpdates, DeferUpdates)
+IMPL_GETTER(std::vector<geode::TextAreaV2::Line>, lines, Lines)
+
+void geode::TextAreaV2::resetMinWidth() {
+    this->setMinWidth(-1);
+}
+
+void geode::TextAreaV2::resetMaxWidth() {
+    this->setMaxWidth(-1);
+}
+
+void geode::TextAreaV2::resetMinHeight() {
+    this->setMinHeight(-1);
+}
+
+void geode::TextAreaV2::resetMaxHeight() {
+    this->setMaxHeight(-1);
+}
+
+void geode::TextAreaV2::resetMinBounds() {
+    this->resetMinWidth();
+    this->resetMinHeight();
+}
+
+void geode::TextAreaV2::resetMaxBounds() {
+    this->resetMaxWidth();
+    this->resetMaxHeight();
+}
+
+void geode::TextAreaV2::setDeferUpdates(const bool defer) {
+    m_deferUpdates = defer;
+
+    // If already deferred, update the container immediately and unschedule the deferred update
+    if (m_deferred) {
+        this->unschedule(schedule_selector(TextAreaV2::updateContainer));
+        
+        this->updateContainer();
+    }
+}
+
+float geode::TextAreaV2::getLineHeight() const {
+    return m_lines.empty() ? 0 : m_lines.front().label->getScaledContentHeight() + m_linePadding;
+}
+
+void geode::TextAreaV2::update() {
+    if (!m_deferred) {
+        if (m_deferUpdates) {
+            m_deferred = true;
+
+            this->scheduleOnce(schedule_selector(TextAreaV2::updateContainer), 0);
+        } else {
+            this->updateContainer();
+        }
+    }
+}
+
+void geode::TextAreaV2::updatePadding() {
+    this->update();
+}
+
+void geode::TextAreaV2::updateContainer(const float dt) {
+    PaddingNode::updatePadding();
+    m_container->removeAllChildren();
+
+    m_deferred = false;
+    m_lines = this->createLines();
+
+    const bool upperBoundWidth = m_maxWidth >= 0;
+    const bool upperBoundHeight = m_maxHeight >= 0;
+    const bool artificialWidth = m_minWidth >= 0 || upperBoundWidth;
+    const bool artificialHeight = m_minHeight >= 0 || upperBoundHeight;
+    const bool inheritedNodeSize = this->getContentSize() == this->getPaddedContainerSize();
+    float height = m_lines.empty() ? 0 : m_lines.back().currentHeight;
+    float width = 0;
+
+    // First determine the container size before manipulating the node anchors and true positions
+    for (const Line& line : m_lines) {
+        width = std::max(width, line.label->getScaledContentWidth());
+    }
+
+    if (artificialWidth) {
+        width = std::max(m_minWidth, width);
+
+        m_container->setContentWidth(upperBoundWidth ? std::min(m_maxWidth, width) : width);
+    } else {
```

Comment:
Everyone else in the Geode team hates me for this so you can ignore this but `else` should be preceded by a newline